Inherit guardian storage

akshay-ap opened the pull request:
This PR changes the design of how the guardians are stored.

Prior to this change, the guardians were stored in a separate contract `GuardianStorage`. The `GuardianStorage` address was then passed as a constructor parameter to `SocialRecoveryModule` which could then make changes to the list on request by the user. So, the recovery functionality was achieved by having two contracts and thus, two storages.

As a side effect, it was initially brought to light that any other module can also modify the list of Guardians in `GuardianStorage` and was not write-restricted only to `SocialRecoveryModule`. Any other enabled module on a Safe could change this list.
See: https://github.com/5afe/CandideWalletContracts/blob/870f82d3c4d286b3fd9241c6479c9472d7d73fc6/contracts/modules/social_recovery/storage/GuardianStorage.sol#L35

The initial argument for accepting this behavior was that any malicious module can:
1- Change ownership of the account
2- Steal all account's funds
3- Execute any arbitrary function of the `SocialRecoveryModule` impersonating the owner
Thus, it did not violate any security assumptions.

Also, allowing the guardian list to be edited by any module allows upgradability of `SocialRecoveryModule` without having to migrate the list of guardians.

But, later auditing partner also reported the same behavior as an issue with the argument:

- There should be a distinction between malicious modules (or any modules that can perform arbitrary operations) and modules that are limited to do these things by their implementation. These "well behaving" modules will effectively gain more privileges than they actually need.
- It opens a new attack vector and violates least privilege principle.

After internal discussion, we concluded that allowing any module to modify the guardians list should not be permitted and this PR reflects the same.

In this PR, `SocialRecoveryModule` inherits `GuardianStorage` so that there is only 1 contract involved in recovery functionality. By making this change and removing `onlyModule` modifier, and other changes in the code (removing `_wallet` as a parameter in some functions) now allows only wallet i.e. `msg.sender` to change the guardians.

This also simplifies the design of the recovery functionality and enforces to use storage space of a single contract i.e. `SocialRecoveryModule`.

Other changes in PR include:
- Fixes to files related to formal verification
- Removing duplicated code which is now not needed
- Remove redundant parameter `_wallet` in some functions

nlordell commented on certora/specs/GuardianStorage.spec:
```diff
@@ -370,6 +370,7 @@ rule addGuardianChangesEntries {
     address toAdd;
```
**As a follow up**, can we merge the FV specs?

akshay-ap replied:
With this PR, we are now running FV on `SocialRecoveryModuleHarness` contract. So, I don't think we need to merge the specs.
https://github.com/5afe/CandideWalletContracts/blob/78bb5d28578f19f41b767f6e03e180624572d6ac/certora/conf/GuardianStorage.conf#L19

nlordell commented on contracts/modules/social_recovery/storage/GuardianStorage.sol:
```diff
@@ -30,24 +30,22 @@ contract GuardianStorage is IGuardianStorage {
+    modifier onlyModule() {
         // solhint-disable-next-line reason-string
-        require(ISafe(payable(_wallet)).isModuleEnabled(msg.sender), "GS: method only callable by an enabled module");
+        require(ISafe(payable(msg.sender)).isModuleEnabled(address(this)), "GS: method only callable by an enabled module");
```
I think the revert message is wrong - it is only callable my an `ISafe`, and if the module is enabled. Is there a reason for this restriction? Shouldn't this just work as `authorized(_wallet)` from before?

akshay-ap replied:
The revert message is indeed incorrect. Fixed here: https://github.com/5afe/CandideWalletContracts/pull/34/commits/81947fe68e75fa56973d87e046392f8da96def61.

This modifier is used to restrict any change in guardians list when module is disabled by the wallet

nlordell replied:
Do we _want_ to restrict changes to the guardian list when the module is disabled?

akshay-ap replied:
I don't have a strong opinion here. But, if I have to choose, then I am in favour of restricting the changes to Guardian list when module is disabled. I treat guardians list as an extended storage to a Safe. I would like to have a tighter connection between state of the Module and Safe such that after disabling module, it becomes impossible to make and changes to guardians. If not a module, this storage should be immutable.

nlordell commented on contracts/modules/social_recovery/storage/GuardianStorage.sol:
```diff
@@ -30,24 +30,22 @@ contract GuardianStorage is IGuardianStorage {
      * @param _guardian The guardian to add.
      */
-    function addGuardianWithThreshold(address _wallet, address _guardian, uint256 _threshold) external onlyModule(_wallet) {
+    function addGuardianWithThreshold(address _guardian, uint256 _threshold) external onlyModule() {
```
Specifically, from my comment in Slack, I would expect this to be:

```suggestion
    function addGuardianWithThreshold(address _wallet, address _guardian, uint256 _threshold) external authorized(_wallet) {
```

This ensures:
1. Consistency with the other `SocialRecoveryModule` methods
2. No breaking ABI change

akshay-ap replied:
I removed this parameter because it is redundant. Also, `authorized` modifier is not needed. I have removed it here: https://github.com/5afe/CandideWalletContracts/pull/34/commits/78bb5d28578f19f41b767f6e03e180624572d6ac

If we want to keep `_wallet` as a parameter just for the consistency, I would revert this change.

nlordell replied:
As discussed internally, we prefer the new interface proposed by @akshay-ap, and we solved the inconsistency with `cancelRecovery`.

nlordell commented on contracts/modules/social_recovery/storage/GuardianStorage.sol:
```diff
@@ -57,39 +55,37 @@ contract GuardianStorage is IGuardianStorage {
      */
-    function changeThreshold(address _wallet, uint256 _threshold) external onlyModule(_wallet) {
-        _changeThreshold(_wallet, _threshold);
+    function changeThreshold(uint256 _threshold) external onlyWhenModuleIsEnabled(){
```
nit: formatting
```suggestion
    function changeThreshold(uint256 _threshold) external onlyWhenModuleIsEnabled() {
```

nlordell commented:
Changes look good. One outstanding question, and before merging can you please update the PR description to include:
- how we changed the contract structure (inheritance vs multiple contracts)
- why we changed it (both the technical reason - to not allow multiple modules to change guardian rules - as well as that this was prompted by the audit and the auditor's comments).

akshay-ap commented:
> Changes look good. One outstanding question, and before merging can you please update the PR description to include:
> 
> * how we changed the contract structure (inheritance vs multiple contracts)
> * why we changed it (both the technical reason - to not allow multiple modules to change guardian rules - as well as that this was prompted by the audit and the auditor's comments).

done

nlordell commented:
Awesome PR description BTW :100:.
